Tutorial - How to add surveys in Next Js (#6952)
PostHog/posthog.com: contents/tutorials/nextjs-surveys.md

---
title: How to set up surveys in Next.js

> date: 2023-10-09

Author: ```suggestion
date: 2023-10-18
```

 
To create a survey with a prebuilt UI, go to the [surveys tab](https://app.posthog.com/surveys), click "New survey," and then set up your survey with the following settings:

1. Add a name (like `my-first-survey`).
2. Set the display mode to `Popover`.
3. Select the `Rating` question type. Set the question title to `How likely are you to recommend us to a friend?`, display type to `number` and scale to `1-10`.
4. Leave the remaining optional properties blank (such as `Targeting` or `Thank you message`).
5. Click "Save as draft" and then on the next screen click "Launch". 

![Popover survey set up](../images/tutorials/nextjs-surveys/create-popover-survey.png)

That's it! Your survey is now live and you should see it in your app. After submitting responses, you can [view results in PostHog](#viewing-results).

![Popover survey in app](../images/tutorials/nextjs-surveys/popover-survey-example.png)

### Option 2: Implement your own survey UI

First, create a Rating survey in PostHog like in option 1 above, except set the display mode to `API`.

![Custom survey set up](../images/tutorials/nextjs-surveys/create-api-survey.png)

> Then, there are three parts to adding code for our custom survey:

Reviewer: ```suggestion
Then, there are four parts to adding code for our custom survey:
```

    const hasInteractedWithSurvey = localStorage.getItem(`hasInteractedWithSurvey_${surveyID}`);
    setShowSurvey(!hasInteractedWithSurvey);
  }, [surveyID]);

  const handleDismiss = () => {
    setShowSurvey(false);
    localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');
    posthog.capture("survey dismissed", {
      $survey_id: surveyID // required
    })
  };

  const handleSubmit = (value) => {
    setShowSurvey(false);
    localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');  
    posthog.capture("survey sent", {
      $survey_id: surveyID, // required
      $survey_response: value // required
    })
  };

  useEffect(() => {
    if (posthog && surveyID && showSurvey) {
      posthog.capture("survey seen", {
        $survey_id: surveyID // required
      })
    }
  }, [showSurvey, surveyID, posthog])

  return (
    <main className={styles.main}>
      <div className="App">
        <h1>This is our Next.js survey tutorial</h1>
        {showSurvey && (
        <Survey
          title={surveyTitle}
          onDismiss={handleDismiss}
          onSubmit={handleSubmit}
        />)}
      </div>
    </main>
  )
}
```

> Our survey is now ready to go! The next step is to submit responses and view your results.

Reviewer: ```suggestion
Our survey is now ready to go! The next step is ship the changes, get responses, and view your results.
```

    }
  }, [posthog]); // posthog may be undefined until it's had a chance to initialize. Hence use it as a dependency for useEffect

  return (
    <main className={styles.main}>
      <div className="App">
        <h1>This is our Next.js survey tutorial</h1>
      </div>
    </main>
  )
}
```

After reloading `localhost:3000`, you should see events appear in your [PostHog events explorer](https://app.posthog.com/events).

![Test events in the PostHog events explorer](../images/tutorials/nextjs-surveys/test-events.png)

## Creating a survey

There are two options for displaying a survey using PostHog:

1. Use PostHog's prebuilt survey UI.
2. Implement your own survey UI.

This tutorial will cover how to implement both options:

### Option 1: Use PostHog's prebuilt survey UI

This is the simplest option. PostHog has a variety of [survey types](/docs/surveys/creating-surveys#question-type) to choose from, and will handles all the display logic and event capture for you.
 
To create a survey with a prebuilt UI, go to the [surveys tab](https://app.posthog.com/surveys) and click "New survey." Then, set up your survey with the following settings:

> 1. Add a name (like `my-first-survey`).
> 2. Set the display mode to `Popover`.
> 3. Select the `Rating` question type. Set the question title to `How likely are you to recommend us to a friend?`, display type to `number` and scale to `1-10`.
> 4. Leave the remaining optional properties blank (such as `Targeting` or `Thank you message`).
> 
> Click "Save as draft" and then on the next screen click "Launch". 

Author: ```suggestion
1. Add a name (like `my-first-survey`).
2. Set the display mode to `Popover`.
3. Select the `Rating` question type. Set the question title to `How likely are you to recommend us to a friend?`, display type to `number` and scale to `1-10`.
4. Leave the remaining optional properties blank (such as `Targeting` or `Thank you message`).
5. Click "Save as draft" and then on the next screen click "Launch". 
```

![Custom survey set up](../images/tutorials/nextjs-surveys/create-api-survey.png)

Then, there are three parts to adding code for our custom survey:

1. Create the survey UI.
2. Add the logic for displaying it.
3. Capture interactions from it.

#### 1. Create the survey UI

We've created a sample survey UI for this tutorial. To use it, create a new file in `./app` folder called `Survey.js` and paste the following code:

```react
// app/Survey.js
function Survey({ title, onDismiss, onSubmit }) {
  const [selectedValue, setSelectedValue] = React.useState(null);

  const handleSelect = (value) => {
    setSelectedValue(value);
  }

  const handleSubmit = () => {
    onSubmit(selectedValue);
  }

  return (
    <div className="survey-popup">
      <h2>{title}</h2>
      <div>
        {[...Array(10)].map((_, i) => (
          <button key={i + 1} onClick={() => handleSelect(i + 1)}>{i + 1}</button>
        ))}
      </div>
      <div>
        <button onClick={onDismiss}>Dismiss</button>
        <button onClick={handleSubmit}>Submit</button>
      </div>
    </div>
  );
}

export default Survey;
```

> Then, add the following CSS styles to your `page.module.css` file:

Author: No its fine, they can leave it

```css
.survey {
  position: fixed;
  bottom: 20px;
  right: 20px;
  width: 400px;
  padding: 20px;
  background-color: #ffffff;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.2);
  border-radius: 5px;
  z-index: 1000;
}

.button {
  margin: 5px;
  padding: 5px;
}
```

> Finally, integrate the component into `page.js`:

Author: Thanks. Will just go with option 1:
> Remove useEffect/usePostHog from earlier, we can rely on autocapture to show that PostHog is working

```react
// app/page.js
'use client'

import { useState } from 'react'
import styles from './page.module.css'
import Survey from './Survey';

export default function Home() {
  const [showSurvey, setShowSurvey] = useState(true);

  const handleDismiss = () => {
    setShowSurvey(false);
    console.log("Survey dismissed!");
  };

  const handleSubmit = (value) => {
    setShowSurvey(false);
    console.log("User submitted:", value);
  };

  return (
    <main className={styles.main}>
      <div className="App">
        <h1>This is our Next.js survey tutorial</h1>
        {showSurvey && (
        <Survey
          title={"Rate our service"}
          onDismiss={handleDismiss}
          onSubmit={handleSubmit}
        />)}
      </div>
    </main>
  )
}
```

This shows a survey popup every time you visit your app's homepage.

> ![Custom survey UI](../images/tutorials/nextjs-surveys/sample-survey-ui.png)

Author: Found the issue, thanks! Fixed now

> #### 2. Add the logic for displaying it.

Author: Hmm not sure I follow, but I'll split it up to following to make it easier to follow:

1. Add UI
2. Add logic for configuring survey
3. Add logic for displaying and hiding it
4. Capture interacitons

The first part of handling our display logic is fetching the survey from PostHog. PostHog keeps track of all active surveys for a user (this is especially helpful if you have set up [custom targeting options](/docs/surveys/creating-surveys#targeting)). 

To fetch the active surveys, we use the `usePostHog` hook to retrieve our PostHog instance. Then, we call `posthog.getActiveMatchingSurveys()` using `useEffect()`:

```react
// app/page.js
'use client'

import { useState, useEffect } from 'react'
import styles from './page.module.css'
import Survey from './Survey';
import { usePostHog } from 'posthog-js/react';

export default function Home() {
  // ... rest of your code ...

  const posthog = usePostHog()
  useEffect(() => {
    posthog.getActiveMatchingSurveys((surveys) => {

>       // TODO: configure the survey

Author: ```suggestion

```

         "descriptionTextColor":"#4b4b52",
         "thankYouMessageHeader":"Thank you for your feedback!",
         "displayThankYouMessage":true
      },
      "start_date":"2023-09-26T09:44:31.844000Z",
      "end_date":null
   }
]
```

We can use this survey object to configure our `Survey` component:

```react
  // ... rest of your code ...

  const [surveyTitle, setSurveyTitle] = useState(false);
  const [surveyID, setSurveyID] = useState(false);

  useEffect(() => {
    posthog.getActiveMatchingSurveys((surveys) => {
      if (surveys.length > 0) {
        const survey = surveys[0];
        setSurveyID(survey.id);
        setSurveyTitle(survey.questions[0].question)
      }
    }); 
  }, [posthog])
  
  // ... rest of your code ...

  return (
    <main className={styles.main}>
      <div className="App">
        <h1>This is our Next.js survey tutorial</h1>
        {showSurvey && (
        <Survey
          title={surveyTitle}
          onDismiss={handleDismiss}
          onSubmit={handleSubmit}
        />)}
      </div>
    </main>
  )
```

Finally, we want to make sure we don't show the survey again to users who have either submitted or dismissed it. 

> We use [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) to store this data. Then, we'll add a check to show the survey based on whether the user has already interacted with it or not:

Reviewer: ```suggestion
We use [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) to store this data. We add a check to show the survey based on whether the user has already interacted with it or not:
```
This sentence doesn't really come after anything.

```react
  // ... rest of your code ...

  const [showSurvey, setShowSurvey] = useState(true);

  useEffect(() => {
    // Check local storage to see if the user has already seen this particular survey
    const hasInteractedWithSurvey = localStorage.getItem(`hasInteractedWithSurvey_${surveyID}`);
    setShowSurvey(!hasInteractedWithSurvey);
  }, [surveyID]);

  const handleDismiss = () => {
    setShowSurvey(false);
    console.log("Survey dismissed!");
    localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');
  };

  const handleSubmit = (value) => {
    setShowSurvey(false);
    console.log("User submitted:", value);
    localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');  

>   };

Reviewer: 1. Could probably combine the setShowSurvey and localStorage set item logic so it doesn't repeat
2. Can get rid of the console.log at this point

Author: > Could probably combine the setShowSurvey and localStorage set item logic so it doesn't repeat

Yeah, if this was real code I'd do that to avoid repetition, but I'll leave as is as I think it's either to follow this way

  // ... rest of your code ...
```

#### 3. Capture interactions from it.

The final step in setting up our survey is capturing interactions. This enables us to analyze the results in PostHog. 

There are 3 events to capture:

1. `"survey shown"`
2. `"survey dismissed"`
3. `"survey sent"` (for responses)

> You can capture these events using `posthog.capture()`:

Reviewer: Should add a note to say make sure posthog is defined before you use it.

Author: PostHog will be defined at this point, since first we need to call `posthog.getActiveMatchingSurveys` in order to show the survey. If its not defined, no survey will be shown

```react
  // ... rest of your code ...

  const handleDismiss = () => {
    setShowSurvey(false);
    console.log("Survey dismissed!");
    localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');
    posthog.capture("survey dismissed", {
      $survey_id: surveyID // required
    })
  };

  const handleSubmit = (value) => {
    setShowSurvey(false);
    console.log("User submitted:", value);
    localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');  
    posthog.capture("survey sent", {
      $survey_id: surveyID, // required
      $survey_response: value // required
    })
  };

  useEffect(() => {
    if (posthog && surveyID && showSurvey) {
      posthog.capture("survey seen", {
        $survey_id: surveyID // required
      })
    }
  }, [showSurvey, surveyID, posthog])

 // ... rest of your code ...
```

Altogether, your code should look like this:

> ```react

Reviewer: Missing the `hasInteractedWithSurvey` check

// app/page.js
'use client'

import { useState, useEffect } from 'react'
import styles from './page.module.css'
import Survey from './Survey';
import { usePostHog } from 'posthog-js/react';

export default function Home() {
  const [showSurvey, setShowSurvey] = useState(true);
  const [surveyTitle, setSurveyTitle] = useState(false);
  const [surveyID, setSurveyID] = useState(false);
  const posthog = usePostHog()

  useEffect(() => {
    posthog.getActiveMatchingSurveys((surveys) => {
      if (surveys.length > 0) {
        const survey = surveys[0];
        setSurveyID(survey.id);
        setSurveyTitle(survey.questions[0].question)
      }
    }); 
  }, [posthog])

  const handleDismiss = () => {
    setShowSurvey(false);

>     console.log("Survey dismissed!");

Reviewer: ```suggestion
```

    localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');
    posthog.capture("survey dismissed", {
      $survey_id: surveyID // required
    })
  };

  const handleSubmit = (value) => {
    setShowSurvey(false);

>     console.log("User submitted:", value);

Reviewer: ```suggestion
```

    localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');  
    posthog.capture("survey sent", {
      $survey_id: surveyID, // required
      $survey_response: value // required
    })
  };

>   useEffect(() => {

Reviewer: This seems to run before the `hasInteractedWithSurvey` check (at least in my implementation), meaning there are "seen survey" events after they submit and without giving them a chance to submit. Need to make sure this works.

Author: Fixed. Need to change default showSurvey state from true to false i.e.

`const [showSurvey, setShowSurvey] = useState(true);` to `  const [showSurvey, setShowSurvey] = useState(false);

    if (posthog && surveyID && showSurvey) {
      posthog.capture("survey seen", {
        $survey_id: surveyID // required
      })
    }
  }, [showSurvey, surveyID, posthog])

  return (
    <main className={styles.main}>
      <div className="App">
        <h1>This is our Next.js survey tutorial</h1>
        {showSurvey && (
        <Survey
          title={surveyTitle}
          onDismiss={handleDismiss}
          onSubmit={handleSubmit}
        />)}
      </div>
    </main>
  )
}
```

> Our survey is now ready to go! The next step is to submit responses and view your results.

Reviewer: Note on removing flickering with bootstrapping?

Author: Unfortunately we'd need to call `posthog.getActiveMatchingSurveys` on the client and wait for the response, so even if we implement bootstrapping, there will be still a "flicker" (its more a delay than a flicker)
